torch/distributed/device_mesh.py

```diff
@@ -570,100 +319,183 @@ def _init_process_groups(
+
+        Returns:
+            DeviceMesh: A :class:`DeviceMesh` object representing the device layout.
+
+        The following program runs on each process/rank in an SPMD manner. In this example, we have 2
+        hosts with 4 GPUs each.
+        A reduction over the first dimension of mesh will reduce across
+        columns (0, 4), .. and (3, 7), a reduction over the second dimension
+        of mesh reduces across rows (0, 1, 2, 3) and (4, 5, 6, 7).
+
+        Example::
+
+            >>> # xdoctest: +SKIP("no rank")
+            >>> from torch.distributed.device_mesh import DeviceMesh
+            >>>
+            >>> # Initialize device mesh as (2, 4) to represent the topology
+            >>> # of cross-host(dim 0), and within-host (dim 1).
+            >>> mesh = DeviceMesh(device_type="cuda", mesh=[[0, 1, 2, 3],[4, 5, 6, 7]])
+        """
+
+        storage: DeviceMeshStorage
+        layouts: list[Layout]
+        mesh_dim_names: Optional[tuple[str, ...]]
+
+        def __init__(
+            self,
+            device_type: str,
+            mesh: Union[torch.Tensor, "ArrayLike"],
+            *,
+            mesh_dim_names: Optional[tuple[str, ...]] = None,
+            backend_override: Optional[
+                tuple[tuple[Optional[str], Optional[C10dBackend.Options]], ...]
+            ] = None,
+            _init_backend: bool = True,
+        ) -> None:
+            if not isinstance(mesh, torch.Tensor):
+                mesh = torch.tensor(mesh, device="cpu", dtype=torch.int)
+            elif not mesh.is_cpu:
+                raise ValueError("...")
+            mesh_dim_names = tuple(mesh_dim_names) if mesh_dim_names else None
+            if backend_override is None:
+                backend_override = ((None, None),) * mesh.ndim
+
+            self.storage = DeviceMeshStorage(device_type, _init_backend=_init_backend)
+            self.layouts = rank_tensor_to_layouts(mesh)
+            self.mesh_dim_names = mesh_dim_names
+
+            for i, layout in enumerate(self.layouts):
+                backend_override_ = backend_override[i]
+                global_override = _mesh_resources.mesh_dim_group_options.get(i, (None, None))
+                if backend_override_ == (None, None):
+                    backend_override_ = global_override
+                elif global_override != (None, None):
+                    raise RuntimeError(
+                        f"Dimension {i} present both in the backend_override argument "
+                        "and via _mesh_resources._set_mesh_dim_group_options"
+                    )
+                self.storage._maybe_create_dim(
+                    layout,
+                    mesh_dim_names[i] if self.mesh_dim_names else None,
+                    backend_override=backend_override_,
+                )
+
+        @staticmethod
+        def from_storage(
+            storage: DeviceMeshStorage,
+            layouts: list[Layout],
```

Comment:
I wonder if this is intended to be a public API to users or only used internally in DeviceMesh? If only supposed to used internally we can make it private? I think a Storage concept + a layout concept might be too hard for end user to understand

Reply:
Yes it could/should be private, you're right